ansible/roles/test/files/ptftests/ip_in_ip_tunnel_test.py

```diff
@@ -0,0 +1,172 @@
+         on all portchannel interfaces to check balance.
+    '''
+    def __init__(self):
+        '''
+        @summary: constructor
+        '''
+        BaseTest.__init__(self)
+        self.test_params = test_params_get()
+
+    def setUp(self):
+        self.server_ip = self.test_params['server_ip']
+        self.active_tor_mac = self.test_params['active_tor_mac']
+        self.standby_tor_mac = self.test_params['standby_tor_mac']
+        self.active_tor_ip = self.test_params['active_tor_ip']
+        self.standby_tor_ip = self.test_params['standby_tor_ip']
+        self.ptf_portchannel_indices = self.test_params['ptf_portchannel_indices']
+        self.indice_to_portchannel = {}
+        for port_channel, indices in self.ptf_portchannel_indices.items():
+            for indice in indices:
+                self.indice_to_portchannel[indice] = port_channel
+
+        self.hash_key_list = self.test_params['hash_key_list']
+        self.dataplane = ptf.dataplane_instance
+
+    def runTest(self):
+        """
+        Entrypoint of test script.
+        """
+        self.send_and_verify_packets()
+
+    def random_ip(self, begin, end):
+        """
+        Generate a random IP from given ip range
+        """
+        length = int(ip_address(end)) - int(ip_address(begin))
+        return str(ip_address(begin) + random.randint(0, length))
+
+    def generate_packet_to_server(self, hash_key):
+        """
+        Generate a packet to server. The value of field in packet is filled with random value according to hash_key
+        """
+        base_src_mac = self.dataplane.get_mac(0, 0)
+        ip_src = self.random_ip(SRC_IP_RANGE[0], SRC_IP_RANGE[1]) if hash_key == 'src-ip' else SRC_IP_RANGE[0]
+        ip_dst = self.server_ip
+        sport = random.randint(1, 65535) if hash_key == 'src-port' else 1234
+        dport = random.randint(1, 65535) if hash_key == 'dst-port' else 80
+        src_mac = (base_src_mac[:-5] + "%02x" % random.randint(0, 255) + ":" + "%02x" % random.randint(0, 255)) if hash_key == 'src-mac' else base_src_mac
+        dst_mac = self.standby_tor_mac
+        vlan_id = random.randint(1, 4094) if hash_key == 'vlan-id' else 0
+        pkt = simple_tcp_packet(pktlen=128 if vlan_id == 0 else 132,
+                            eth_dst=dst_mac,
+                            eth_src=src_mac,
+                            dl_vlan_enable=False if vlan_id == 0 else True,
+                            vlan_vid=vlan_id,
+                            vlan_pcp=0,
+                            ip_src=ip_src,
+                            ip_dst=ip_dst,
+                            tcp_sport=sport,
+                            tcp_dport=dport,
+                            ip_ttl=64)
+        return pkt
+
+    def generate_expected_packet(self, inner_pkt):
+        """
+        Generate ip_in_ip packet for verifying.
+        """
+        inner_pkt = inner_pkt.copy()
+        inner_pkt.ttl = inner_pkt.ttl - 1
+        """
```

Comment:
The comment here should be removed, right?

Reply:
Thanks. Updated.